plugins/by-name/typescript-tools/default.nix

```diff
@@ -157,38 +158,117 @@ in
-      in
-      ''
-        require('typescript-tools').setup(${helpers.toLuaObject options})
+
+  settingsExample = {
+    handlers = {
+      "textDocument/publishDiagnostics" = ''
+        api.filter_diagnostics(
+          { 80006 }
+        )
+      '';
+    };
+    settings = {
+      tsserver_plugins = [
+        "@styled/typescript-styled-plugin"
+      ];
+      tsserver_file_preferences.__raw = ''
+        function(ft)
+          return {
+            includeInlayParameterNameHints = "all",
+            includeCompletionsForModuleExports = true,
+            quotePreference = "auto",
+          }
+        end
+      '';
+      tsserver_format_options.__raw = ''
+        function(ft)
+          return {
+            allowIncompleteCompletions = false,
+            allowRenameOfImportPath = false,
+          }
+        end
       '';
+    };
   };
+
+  # NOTE: call setup in lsp config
+  callSetup = false;
+  extraConfig =
+    cfg:
+    let
+      definedOpts = lib.filter (opt: lib.hasAttrByPath opt cfg.settings) [
+        [ "separateDiagnosticServer" ]
+        [ "publishDiagnosticOn" ]
+        [ "exposeAsCodeAction" ]
+        [ "tsserverPath" ]
+        [ "tsserverPlugins" ]
+        [ "tsserverMaxMemory" ]
+        [ "tsserverFormatOptions" ]
+        [ "tsserverFilePreferences" ]
+        [ "tsserverLocale" ]
+        [ "completeFunctionCalls" ]
+        [ "includeCompletionsWithInsertText" ]
+        [ "codeLens" ]
+        [ "disableMemberCodeLens" ]
+        [
+          "jsxCloseTag"
+          "enable"
+        ]
+        [
+          "jsxCloseTag"
+          "filetypes"
+        ]
+      ];
+    in
+    {
+      plugins.lsp.postConfig =
+        let
+          # TODO:: introduced 10-22-2024: remove after 24.11
+          renamedSettings = lib.foldl' (
+            acc: optPath:
+            let
+              snakeCasePath = builtins.map lib.nixvim.toSnakeCase optPath;
+              optValue = lib.getAttrFromPath optPath cfg.settings;
+            in
+            lib.recursiveUpdate acc (lib.setAttrByPath snakeCasePath optValue)
+          ) { } definedOpts;
+
+          filterAttrsRecursivePath =
+            predicate: set: path:
+            lib.listToAttrs (
+              lib.concatMap (
+                name:
+                let
+                  path' = path ++ [ name ];
+                  v = set.${name};
+                in
+                lib.optional (predicate path' v) {
+                  inherit name;
+                  value = if lib.isAttrs v then filterAttrsRecursivePath predicate v path' else v;
+                }
+              ) (lib.attrNames set)
+            );
+
+          setupOptions =
+            filterAttrsRecursivePath (
+              path: value: !builtins.elem path definedOpts
+              # TODO: Maybe also remove if value is an attrset whose only children are in definedOpts?
+              # Although, that isn't important because toLuaObject will filter out empty attrsets anyway...
```

Comment:
```suggestion
```

Keep the comments if you want, but when I first wrote them I only really intended them for the review :slightly_smiling_face:

Reply:
lol sorry i'm on auto pilot this morning